storage: support AWS KMS server side encryption

adityamaru opened the pull request:
S3 URIs now support two additional params to provide server side
encryption:

AWS_SERVER_ENC_MODE: can be set to either AES256 or aws:kms

AWS_SERVER_KMS_ID: if the above is set to aws:kms then this is the ID of
the CMK to be used

Closes: #31612

Release note (api change): `AWS_SERVER_ENC_MODE` and `AWS_SERVER_KMS_ID`
can now be specified as a param in all s3 URIs which write to the store.
(egs: Changefeed, BACKUP)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55834)
<!-- Reviewable:end -->

adityamaru commented:
Trying to see if there's an easy way to grab object metadata and check that server-side encryption is activated in the unit test, but I have manually verified it at the following buckets cc: @mwang1026 

`s3://michael-crdb-test/backups/test-sse-kms` and `s3://michael-crdb-test/backups/test-sse-256`

<img width="771" alt="Screen Shot 2020-10-21 at 5 27 19 PM" src="https://user-images.githubusercontent.com/13837382/96789336-f7fbf580-13c2-11eb-8d2e-e700ffaad5c2.png">
<img width="910" alt="Screen Shot 2020-10-21 at 5 27 07 PM" src="https://user-images.githubusercontent.com/13837382/96789338-f92d2280-13c2-11eb-883e-dcadcf15c718.png">

pbardea reviewed (approved):
LGTM with some nits, but I'm not super familiar with this area yet, so would appreciate another set of eyes. :)

adityamaru committed:
storage: support AWS KMS server side encryption

S3 URIs now support two additional params to provide server side
encryption:

AWS_SERVER_ENC_MODE: can be set to either AES256 or aws:kms

AWS_SERVER_KMS_ID: if the above is set to aws:kms then this is the
ID of the CMK to be used

Release note (api change): `AWS_SERVER_ENC_MODE` and `AWS_SERVER_KMS_ID`
can now be specified as a param in all s3 URIs which write to the store.
(egs: Changefeed, BACKUP)

adityamaru commented on pkg/storage/cloudimpl/s3_storage.go:
done

adityamaru commented on pkg/roachpb/api.proto:
```diff
@@ -1328,6 +1328,8 @@ message ExternalStorage {
     string endpoint = 6;
     string region = 7;
     string auth = 8;
+    string server_enc_mode  = 9 [(gogoproto.customname) = "ServerEncMode"];
```
good point, done.

adityamaru commented on pkg/storage/cloudimpl/cloudimpltests/s3_storage_test.go:
```diff
@@ -93,6 +113,74 @@ func TestPutS3(t *testing.T) {
+		credentialsProvider := credentials.SharedCredentialsProvider{}
+		_, err := credentialsProvider.Retrieve()
+		if err != nil {
+			skip.IgnoreLint(t, err)
```
done

adityamaru commented:
@dt can I get a quick review on this, Paul's done a round but would like another set of 👀

dt commented on pkg/storage/cloudimpl/s3_storage.go:
does this sort of validation feel like it might belong in `MakeS3Storage` instead? seems like `Put` is kinda late to decide the config is invalid?

adityamaru replied:
done, moved this, and checking that the mode is one we support, to MakeS3Storage.

adityamaru commented:
TFTR!

bors r=pbardea,dt
